packages/actor-extract-links-extract-tree/lib/solver.ts

```diff
@@ -0,0 +1,395 @@
+ */
+export function resolveAFilterTerm(expression: Algebra.Expression,
+  operator: SparqlRelationOperator,
+  linksOperator: LinkOperator[],
+  variable: Variable):
+  ISolverExpression | Error {
+  let rawValue: string | undefined;
+  let valueType: SparqlOperandDataTypes | undefined;
+  let valueAsNumber: number | undefined;
+  let hasVariable = false;
+
+  // Find the constituant element of the solver expression
+  for (const arg of expression.args) {
+    if ('term' in arg && arg.term.termType === 'Variable') {
+      // Check if the expression has the same variable as the one the solver try to resolved
+      if (arg.term.value !== variable) {
+        return new MissMatchVariableError(`the variable ${arg.term.value} is in the filter whereas we are looking for the varibale ${variable}`);
+      }
+      hasVariable = true;
+    } else if ('term' in arg && arg.term.termType === 'Literal') {
+      rawValue = arg.term.value;
+      valueType = SparqlOperandDataTypesReversed.get(arg.term.datatype.value);
+      if (valueType) {
+        valueAsNumber = castSparqlRdfTermIntoNumber(rawValue!, valueType);
+        if (!valueAsNumber) {
+          return new UnsupportedDataTypeError(`we do not support the datatype "${valueType}" in the solver for the moment`);
+        }
+      } else {
+        return new UnsupportedDataTypeError(`The datatype "${valueType}" is not supported by the SPARQL 1.1 Query Language W3C recommandation`);
+      }
+    }
+  }
+  // Return if a fully form solver expression can be created
+  if (hasVariable && rawValue && valueType && valueAsNumber) {
+    return {
+      variable,
+      rawValue,
+      valueType,
+      valueAsNumber,
+      operator,
+      chainOperator: linksOperator,
+    };
+  }
+  const missingTerm = [];
+  if (!hasVariable) {
+    missingTerm.push('Variable');
+  }
+  if (!rawValue) {
+    missingTerm.push('Litteral');
+  }
+
+  return new MisformatedFilterTermError(`the filter expressions doesn't have the term ${missingTerm.toString()}`);
+}
+
+/**
+ * Recursively traverse the filter expression and calculate the domain until it get to the current expression.
+ * It will thrown an error if the expression is badly formated or if it's impossible to get the solution range.
+ * @param {Algebra.Expression} filterExpression - The current filter expression that we are traversing
+ * @param {SolutionDomain} domain - The current resultant solution domain
+ * @param {LogicOperator} logicOperator - The current logic operator that we have to apply to the boolean expression
+ * @param {Variable} variable - The variable targeted inside the filter expression
+ * @param {boolean} notExpression
+ * @returns {SolutionDomain} The solution domain of the whole expression
+ */
+export function recursifResolve(
+  filterExpression: Algebra.Expression,
+  domain: SolutionDomain,
+  logicOperator: LogicOperator | undefined,
+  variable: Variable,
+  notExpression: boolean,
```

Comment:
Why is this needed? Whether or not we are evaluation a `not` should only depend on the current expression, right?

Reply:
Not if the not operator is applied to a whole parenthesis like `!(x>5 || x =2  || x = -1)`

Reply:
That would be a nested expression, so we can solve that recursively.

Reply:
I felt it was easier to distribute the not expression, then to inverse the whole domain, but I will check.

Reply:
> I see now that because I didn't wanted to implement an inverse domain function

But you already have an `inverse` implementation, so it should be pretty straightforward I guess?